File: daily_summaries.R
library(data.table)
library(lubridate)

source("NG_FWI.r")
source("util.r")


smooth_5pt <- function(source) {
  #binomial smoother  ... specifically for the 24 hour day
  #1pt = 1
  #3pt = (1 2 1) = 4
  #5pt = (1 4 6 4 1) = 16
  #7pt = (1 6 15 20 15 6 1) = 64
  
  cap <- length(source) #normally 24 edge cases on data set input though
  
  dest <- numeric(cap)
  
  dest[1] <- source[1]
  dest[cap] <- source[cap]
  
  miss <- 0
  for (i in 1:3){
    if (source[i] < -90.0){
      miss <- miss + 1
    }
  }
  if (miss == 0){
    dest[2] <- (0.25 * source[1]) + (0.5 * source[2]) + (0.25 * source[3])
  }
  else {
    dest[2] <- source[2]
  }
  
  
  for (i in 3:(cap-2)){
    miss <- 0
    for (j in (i-2):(i+2)){
      
      if (source[j] < -90.0){
        miss <- miss + 1
      }
    }
    if (miss == 0){
      dest[i] <- (1.0/16.0 * source[i - 2]) + (4.0/16.0 * source[i - 1]) + (6.0/16.0 * source[i]) + (4.0/16.0 * source[i + 1]) + (1.0/16.0 * source[i + 2])
    }
    else {
      dest[i] <- source[i]
    }
  }
  
  
  miss <- 0 
  for (i in (cap-2):cap){
    if (source[i] < -90.0){
      miss <- miss + 1
    }
  }
  if (miss == 0){
    dest[cap - 1] <- (0.25 * source[cap - 2]) + (0.5 * source[cap - 1]) + (0.25 * source[cap])
  }
  else {
    dest[cap - 1] <- source[cap - 1]
  }
  
  return(dest)
}


#smooth_7pt <- function(source) {
#binomial smoother  ... specifically for the 24 hour day
#1pt = 1
#3pt = (1 2 1) = 4
#5pt = (1 4 6 4 1) = 16
#7pt = (1 6 15 20 15 6 1) = 64


#  dest <- numeric(24)

#  dest[1] <- source[1]
#  dest[24] <- source[24]

#  miss <- 0
#  for (i in 1:3){
#    if (source[i] < -90.0){
#      miss <- miss + 1
#    }
#  }
#  if (miss == 0){
#    dest[2] <- (0.25 * source[1]) + (0.5 * source[2]) + (0.25 * source[3])
#  }
#  else {
#    dest[2] <- source[2]
#  }


#  miss <- 0
#  for (i in 1:5){
#    if (source[i] < -90.0){
#      miss <- miss + 1
#    }
#  }
#  if (miss == 0){
#    dest[3] <- (1.0/16.0 * source[1]) + (4.0/16.0 * source[2]) + (6.0/16.0 * source[3]) + (4.0/16.0 * source[4]) + (1.0/16.0 * source[5])
#  }
#  else {
#    dest[3] <- source[3]
#  }


#  for (i in 4:21){
#    miss <- 0
#    for (j in (i-3):(i+3)){
#      if (source[j] < -90.0){
#        miss <- miss + 1
#      }
#    }
#    if (miss == 0){
#      dest[i] <- (1.0/64.0 * source[i - 3]) + (6.0/64.0 * source[i - 2]) + (15.0/64.0 * source[i - 1]) + (20.0/64.0 * source[i]) + (15.0/64.0 * source[i + 1]) + (6.0/64.0 * source[i + 2]) + (1.0/64.0 * source[i + 3])
#    }
#    else {
#      dest[i] <- source[i]
#    }
#  }


#  miss <- 0
#  for (i in 20:24){
#    if (source[i] < -90.0){
#      miss <- miss + 1
#    }
#  }
#  if (miss == 0){
#    dest[22] <- (1.0/16.0 * source[20]) + (4.0/16.0 * source[21]) + (6.0/16.0 * source[22]) + (4.0/16.0 * source[23]) + (1.0/16.0 * source[24])
#  }
#  else {
#    dest[3] <- source[3]
#  }


#  miss <- 0 
#  for (i in 22:24){
#    if (source[i] < -90.0){
#      miss <- miss + 1
#    }
#  }
#  if (miss == 0){
#    dest[23] <- (0.25 * source[22]) + (0.5 * source[23]) + (0.25 * source[24])
#  }
#  else {
#    dest[23] <- source[23]
#  }


#  return(dest)
#}

#' Calculate a pseudo-date that changes not at midnight but forward at another hour
#'
#' @param   yr        year
#' @param   mon       month number
#' @param   day       day of month
#' @param   hr        hour of day
#' @param   reset_hr  the new boundary hour instead of midnight (default 5)
#' @return            pseudo-date including year and ordinal day of the form "YYYY-D"
pseudo_date <- function(yr, mon, day, hr, reset_hr = 5) {
  if (hr < reset_hr) {
    adjusted_jd <- julian(mon, day) - 1
  } else {
    adjusted_jd <- julian(mon, day)
  }

  if (adjusted_jd == 0) {  # where Jan 1 shifts to 0, bump it to end of previous year
    adjusted_jd <- julian(12, 31)
    adjusted_yr <- yr - 1
  } else {
    adjusted_yr <- yr
  }

  return(sprintf("%d-%d", adjusted_yr, adjusted_jd))
}

#' Calculate hourly FWI indices from hourly weather stream.
#'
#' @param     hourly_FWI    hourly FWI dataframe (output of hFWI())
#' @param     reset_hr      new boundary to define day to summarize (default 5)
#' @param     silent        suppresses informative print statements (default False)
#' @param     round_out     decimals to truncate output to, None for none (default 4)
#' @return                  daily summary of peak FWI conditions
generate_daily_summaries <- function(hourly_FWI, reset_hr = 5,
  silent = FALSE, round_out = 4) {
  wasDf <- is.data.frame(hourly_FWI)
  hourly_data <- copy(hourly_FWI)
  if (wasDf) {
    setDT(hourly_data)
  } else if (!is.data.table(hourly_data)) {
    stop("Input hourly FWI needs to be a data.frame or data.table!")
  }

  Spread_Threshold_ISI <- 5.0

  # check for "id" column
  if ("id" %in% names(hourly_data)) {
    had_stn <- TRUE
  } else {
    if (uniqueN(hourly_data, by = c("yr", "lat", "long")) == 1) {
      hourly_data[, id := "stn"]
      had_stn <- FALSE
    } else {
      stop("Missing 'id' column with multiple years and locations in data")
    }
  }

  results <- NULL

  for (stn in unique(hourly_data[, id])) {
    if (!silent) {
      print(paste("Summarizing", stn, "to daily"))
    }
    by_stn <- hourly_data[id == stn]
    by_stn[, pseudo_DATE := Vectorize(pseudo_date)(yr, mon, day, hr, reset_hr)]

    for (p_date in unique(by_stn[, pseudo_DATE])) {
      by_date <- by_stn[pseudo_DATE == p_date]

      # if this pseudo-date doesn't have more than 12 hours, skip
      if (nrow(by_date) <= 12) {
        next
      }

      # find daily peak burn times
      by_date <- by_date[, ws_smooth := smooth_5pt(ws)]
      by_date <- by_date[, isi_smooth := initial_spread_index(ws_smooth, ffmc)]

      max_ffmc <- by_date[, max(ffmc)]
      if (max_ffmc < 85.0) {
        peak_time <- 13  # 12 hours into pseudo-date
      } else {
        peak_time <- by_date[, which.max(isi_smooth)]
      }

      # calculate some extra variables before creating datatable all at once
      if (by_date[1, julian(mon, day)] < DATE_GRASS) {
        standing <- FALSE
      } else {
        standing <- TRUE
      }
      # load to format sunrise and sunset as hh:mm from decimal hours later
      sr <- by_date[peak_time, sunrise]
      ss <- by_date[peak_time, sunset]

      # find the rest of the values at peak
      daily_report <- data.table(id = by_date[1, id],
        yr = by_date[1, yr],
        mon = by_date[1, mon],
        day = by_date[1, day],
        sunrise = sprintf("%02d:%02d", trunc(sr), trunc(60 * (sr - trunc(sr)))),
        sunset = sprintf("%02d:%02d", trunc(ss), trunc(60 * (ss - trunc(ss)))),
        peak_hr = by_date[peak_time, hr],
        duration = by_date[, sum(isi_smooth > Spread_Threshold_ISI)],
        ffmc = by_date[peak_time, ffmc],
        dmc = by_date[peak_time, dmc],
        dc = by_date[peak_time, dc],
        isi = by_date[peak_time, isi],
        bui = by_date[peak_time, bui],
        fwi = by_date[peak_time, fwi],
        dsr = by_date[peak_time, dsr],
        gfmc = by_date[peak_time, gfmc],
        gsi = by_date[peak_time, gsi],
        gfwi = by_date[peak_time, gfwi],
        ws_smooth = by_date[peak_time, ws_smooth],
        isi_smooth = by_date[peak_time, isi_smooth],
        gsi_smooth = by_date[peak_time,
          grass_spread_index(ws_smooth, gfmc, percent_cured, standing)])

      results <- rbind(results, daily_report)
    }
  }

  if (!had_stn) {
    results <- results[, -"id"]
  }

  # format decimal places of output columns
  if (!is.na(round_out)) {
    outcols <- c("ffmc", "dmc", "dc", "isi", "bui", "fwi", "dsr",
      "gfmc", "gsi", "gfwi", "ws_smooth", "isi_smooth", "gsi_smooth")
    set(results, j = outcols, value = round(results[, ..outcols], round_out))
  }

  if (wasDf) {
    setDF(results)
  }

  return(results)
}

# run generate_daily_summaries by command line via Rscript, requires input and output csv
# optional args: reset_hr, silent, round_out
if ("--args" %in% commandArgs() && sys.nframe() == 0) {
  args <- commandArgs(trailingOnly = TRUE)
  if (length(args) < 2) {
    stop("at least 2 arguments required: input csv, output csv")
  }
  input <- args[1]
  output <- args[2]
  # load optional arguments if provided, or set to default
  if (length(args) >= 3) reset_hr <- args[3]
  else reset_hr <- 5
  if (length(args) >= 4) silent <- as.logical(args[4])
  else silent <- FALSE
  if (length(args) >= 5) round_out <- args[5]
  else round_out <- 4
  if (length(args) >= 6) warning("Too many input arguments provided, some unused")

  df_in <- read.csv(input)
  df_out <- generate_daily_summaries(df_in, reset_hr, silent, round_out)
  write.csv(df_out, output, row.names = FALSE)
}

File: daily_summaries.py
import logging
import os.path
import sys
import pandas as pd
import NG_FWI
import util

logger = logging.getLogger("cffdrs")
logger.setLevel(logging.WARNING)


def smooth_5pt(source):
  #binomial smoother  ... specifically for the 24 hour day
  #1pt = 1
  #3pt = (1 2 1) = 4
  #5pt = (1 4 6 4 1) = 16
  #7pt = (1 6 15 20 15 6 1) = 64


  cap = source.__len__()
  dest = [None]*cap
  

  dest[0] = source.iloc[0]
  dest[cap -1] = source.iloc[cap -1]
  
  miss = 0
  for i in range(0,2 +1):
    if source.iloc[i] < -90.0:
      miss = miss + 1
  if miss == 0:
    dest[1] = (0.25 * source.iloc[0]) + (0.5 * source.iloc[1]) + (0.25 * source.iloc[2])
  else:
    dest[1] = source.iloc[1]
  
  for i in range(2,(cap-3)+1):
    miss = 0
    for j in range((i-2),(i+2)+1):
      if source.iloc[j] < -90.0:
        miss = miss + 1
    if miss == 0:
      dest[i] = (1.0/16.0 * source.iloc[i - 2]) + (4.0/16.0 * source.iloc[i - 1]) + (6.0/16.0 * source.iloc[i]) + (4.0/16.0 * source.iloc[i + 1]) + (1.0/16.0 * source.iloc[i + 2])
    else:
      dest[i] = source.iloc[i]
  
  miss = 0
  for i in range((cap-3),(cap-1)+1):
    if source.iloc[i] < -90.0:
       miss - miss + 1
  if miss == 0:
    dest[cap-2] = (0.25 * source.iloc[cap - 3]) + (0.5 * source.iloc[cap - 2]) + (0.25 * source.iloc[cap-1])
  else:
    dest[cap-2] = source.iloc[cap-2]
  return dest
  

#this function calculates a fake date based on 5am-5am instead of midnight to midnight
#used for generating daily summaries to make data look at 5am-5am instead
#output form is "year-julian_day", with the julian day rounded back if the time is before 5am
#for Jan 1st where the julian day rollback would make it zero it gets bumped to the last day of the previous year
def pseudo_date(year, month, day, hour):
  
  adjusted_jd = None
  adjusted_year = None
  if (hour >= 5):
    adjusted_jd = util.julian(month, day)
  else:
    adjusted_jd = util.julian(month, day) - 1
  
  if adjusted_jd == 0:
    adjusted_year = year - 1
    adjusted_jd = util.julian(12,31)
  else:
    adjusted_year = year
    
  out = "{}-{}".format(adjusted_year, adjusted_jd)
  return out


def generate_daily_summaries(hourly_data):
  #note: need to account for spill over inlast day after pseudo_date calc where there is not 24 hours in the data
  
  Spread_Threshold_ISI = 5.0
  
  
  cols = ["wstind","yr","mon","day","peak_time","duration","wind_speed_smoothed","peak_isi_smoothed","peak_gsi_smoothed","ffmc","dmc","dc","isi","bui","fwi","dsr","gfmc","gsi","gfwi","sunrise","sunset"]
  results = pd.DataFrame(columns=cols)
  
  for stn in hourly_data["id"].unique():
    print("Summarizing " + stn + " to daily")
    by_stn = hourly_data.loc[hourly_data["id"] == stn]
    by_stn["pseudo_DATE"] =  by_stn.apply(
            lambda row: pseudo_date(
                row["yr"], row["mon"], row["day"], row["hr"]
            ),
            axis=1,
        )
    
    
    for p_date in by_stn["pseudo_DATE"].unique():
      by_date = by_stn.loc[by_stn["pseudo_DATE"] == p_date]

      # if this day doesn't go up to hour 17, skip
      if ((by_date.reset_index()).index[by_date["hr"] == 17].tolist().__len__() == 0):
        continue

      peak_time_traditional_spot = (by_date.reset_index()).index[by_date["hr"] == 17].tolist()[0]
      
      
      
      peak_time = -1
      duration = 0
      wind_smooth = smooth_5pt(by_date["ws"])
      peak_isi_smooth = -1
      peak_gsi_smooth = -1
      max_ffmc = 0
      ffmc = 0
      gfmc = 0
      dmc = 0
      dc = 0
      isi = 0
      gsi = 0
      bui = 0
      fwi = 0
      gfwi = 0
      dsr = 0

      
      

      for i in range(0,by_date.shape[0]):
        smooth_isi = 0
        if (wind_smooth[i] > -90.0) and (by_date.iloc[i]["ffmc"] > -90.0):
          smooth_isi = NG_FWI.initial_spread_index(wind_smooth[i], by_date.iloc[i]["ffmc"])
        else:
          smooth_isi = -98.9
      
        if smooth_isi > peak_isi_smooth:
          peak_time = i
          peak_isi_smooth = smooth_isi
        if by_date.iloc[i]["ffmc"] > max_ffmc:
          max_ffmc = by_date.iloc[i]["ffmc"]
        if smooth_isi > Spread_Threshold_ISI:
          duration = duration + 1
       
      if (smooth_isi < 5) and (duration == 24):
         duration = 0
      
      if max_ffmc < 85.0:
        peak_time = peak_time_traditional_spot
    
      ffmc = by_date.iloc[peak_time]["ffmc"]
      dmc = by_date.iloc[peak_time]["dmc"]
      dc = by_date.iloc[peak_time]["dc"]
      isi = by_date.iloc[peak_time]["isi"]
      bui = by_date.iloc[peak_time]["bui"]
      fwi = by_date.iloc[peak_time]["fwi"]
      dsr = by_date.iloc[peak_time]["dsr"]
      smooth_ws_peak = wind_smooth[peak_time]
      
      gfmc = by_date.iloc[peak_time]["gfmc"]
      gsi = by_date.iloc[peak_time]["gsi"]
      gfwi = by_date.iloc[peak_time]["gfwi"]

    


      pick_year = by_date["yr"].unique()
      if pick_year.shape[0] > 1:
        pick_year = pick_year[0]
      else:
         pick_year = pick_year[0]
      pick_month = by_date["mon"].unique()
      if pick_month.shape[0] > 1:
        pick_month = pick_month[0]
      else:
         pick_month = pick_month[0]
      pick_day = by_date["day"].unique()
      if pick_day.shape[0] > 1:
        pick_day = pick_day[0]
      else:
        pick_day = pick_day[0]


      standing = True
      if (util.julian(pick_month, pick_day) < NG_FWI.DATE_GRASS):
        standing = False
      peak_gsi_smooth = NG_FWI.grass_spread_index(smooth_ws_peak, gfmc, by_date.iloc[peak_time]["percent_cured"], standing)

      sunrise_val = by_date.iloc[peak_time]["sunrise"]
      sunset_val = by_date.iloc[peak_time]["sunset"]
    
      peak_time = by_date.iloc[peak_time]["hr"]

      sunrise_formated = "{}:{}".format(int(sunrise_val),int(60*(sunrise_val-int(sunrise_val))))
      sunset_formated = "{}:{}".format(int(sunset_val),int(60*(sunset_val-int(sunset_val))))


      daily_report_line = [by_date["id"].unique()[0], pick_year, pick_month, pick_day,
        peak_time, duration, smooth_ws_peak, peak_isi_smooth, peak_gsi_smooth,
        ffmc, dmc, dc, isi, bui, fwi, dsr, gfmc, gsi, gfwi,
        sunrise_formated, sunset_formated]
      
      daily_report = pd.DataFrame(columns=cols)
      daily_report.loc[0] = daily_report_line

      
      results = pd.concat([results, daily_report])

  
  results = results[["wstind","yr","mon","day","peak_time","duration","wind_speed_smoothed","peak_isi_smoothed","peak_gsi_smoothed","ffmc","dmc","dc","isi","bui","fwi","dsr","gfmc","gsi","gfwi","sunrise","sunset"]]

  return results


if "__main__" == __name__:
    args = sys.argv[1:]
    if len(args) != 2:
        logger.fatal(f"{sys.argv[0]} arguments must be: <input file> <output file>\n")
        sys.exit(1)
    infile = args[0]
    outfile = args[1]
    hourly_data = pd.read_csv(infile)
    summaries = generate_daily_summaries(hourly_data)
    util.save_csv(summaries, outfile)
